extensions/guacamole-auth-quickconnect/src/main/java/org/apache/guacamole/auth/quickconnect/QuickConnectDirectory.java

```diff
@@ -0,0 +1,127 @@
+import org.apache.guacamole.net.auth.simple.SimpleDirectory;
+import org.apache.guacamole.net.auth.Connection;
+import org.apache.guacamole.protocol.GuacamoleConfiguration;
+
+/**
+ * Implementation of a directory to stored Connection objects
+ * completely in-memory.
+ */
+public class QuickConnectDirectory extends SimpleDirectory<Connection> {
+
+    /**
+     * The unique identifier of the root connection group.
+     */
+    private static final String ROOT_IDENTIFIER = "ROOT";
+
+    /**
+     * The connections to store.
+     */
+    private final Map<String, Connection> connections =
+            new HashMap<String, Connection>();
+
+    /**
+     * The root connection group for this directory.
+     */
+    private final QuickConnectionGroup rootGroup;
+
+    /**
+     * The internal counter for connection IDs.
+     */
+    private AtomicInteger connectionId;
+
+    /**
+     * Creates a new QuickConnectDirectory with the default
+     * empty Map for Connection objects, and the specified
+     * ConnectionGroup at the root of the directory.
+     *
+     * @param rootGroup
+     *     A group that should be at the root of this directory.
+     */
+    public QuickConnectDirectory(ConnectionGroup rootGroup) {
+        this.rootGroup = (QuickConnectionGroup)rootGroup;
+        this.connectionId = new AtomicInteger();
+        super.setObjects(this.connections);
+    }
+
+    /**
+     * Returns the current connection identifier counter and
+     * then increments it.
+     *
+     * @return
+     *     An int representing the next available connection
+     *     identifier to be used when adding connections.
+     */
+    private int getNextConnectionID() {
+        return connectionId.getAndIncrement();
+    }
+
+    @Override
+    public void add(Connection connection) throws GuacamoleException {
+        connections.put(connection.getIdentifier(), connection);
+    }
+
+    /**
+     * Create a SimpleConnection object from a GuacamoleConfiguration,
+     * obtain an identifier, and place it on the tree, returning the
+     * identifier value of the new connection.
+     *
+     * @param config
+     *     The GuacamoleConfiguration to use to create the
+     *     SimpleConnection object.
+     *
+     * @return
+     *     The identifier of the connection created in the directory.
+     *
+     * @throws GuacamoleException
+     *     If an error occurs adding the object to the tree.
+     */
+    public String create(GuacamoleConfiguration config) throws GuacamoleException {
+
+        // Get the next connection identifier.
+        String connectionId = Integer.toString(getNextConnectionID());
+
+        // Generate a name for the configuration.
+        String name = QCParser.getName(config);
+
+        // Create a new connection and set the parent identifier.
+        Connection connection = new SimpleConnection(name, connectionId, config);
+        connection.setParentIdentifier(ROOT_IDENTIFIER);
```

Comment:
The `QuickConnectUserContext` uses `DEFAULT_ROOT_CONNECTION_GROUP`, which is not guaranteed to match `ROOT_IDENTIFIER` here. As the identifier of the root group is in fact `DEFAULT_ROOT_CONNECTION_GROUP`, that should be the value used here.

Reply:
So, do I need to inject or pass in the `UserContext` object, here?  The `DEFAULT_ROOT_CONNECTION_GROUP` string is protected in the class, so I can't just use it...

Reply:
Done.